expand.go

```diff
@@ -22,16 +22,64 @@ package config
 
 import (
 	"bytes"
+	"fmt"
+	"strings"
 
 	"golang.org/x/text/transform"
 )
 
+const (
+	_envSeparator = ":"
+	_emptyDefault = `""`
+)
+
+// A LookupFunc behaves like os.LookupEnv: it uses the supplied string as a
+// key into some key-value store and returns the value and whether the key was
+// present.
+type LookupFunc = func(string) (string, bool)
+
+// Given a function with the same signature as os.LookupEnv, return a function
+// that expands expressions of the form ${ENV_VAR:default_value}.
+func replace(lookUp LookupFunc) func(in string) (string, error) {
+	return func(in string) (string, error) {
+		sep := strings.Index(in, _envSeparator)
+		var key string
+		var def string
+
+		if sep == -1 {
+			// separator missing - everything is the key ${KEY}
+			key = in
+		} else {
+			// ${KEY:DEFAULT}
+			key = in[:sep]
+			def = in[sep+1:]
+		}
+
+		if envVal, ok := lookUp(key); ok {
+			return envVal, nil
+		}
+
+		if def == "" {
+			return "", fmt.Errorf(`default is empty for %q (use "" for empty string)`, key)
```

Comment:
What if the key isn't found and it is of form `{$KEY}`? In this case, `def` will be `""`, but we will say `"default is empty ..."`. The default might be intentionally empty. Would it be better to say `$KEY could not be determined from the environment, or something similar?

Reply:
Possibly. This whole function was copied verbatim from a now-deleted file - I'd prefer to keep it exactly as-is right now.

The `expand.go` file is a bit of a mess, and it'll need a round of refactoring just so the project passes lint. We can address this stuff during that pass.

Reply:
Sounds good!